Blish HUD/Controls/ColorPicker.cs

```diff
@@ -10,106 +10,120 @@ namespace Blish_HUD.Controls {
+
+                    associatedColorBox = value;
+                    associatedColorBox.Selected = true;
+                    this.SelectedColor = this.AssociatedColorBox.Color;
+                }
+            }
+        }
+
+        public ColorPicker() : base() {
+            this.Colors = new ObservableCollection<Gw2Sharp.WebApi.V2.Models.Color>();
+            this.Colors.CollectionChanged += Colors_CollectionChanged;
+
+            this.ContentRegion = new Rectangle(COLOR_PADDING, COLOR_PADDING, (this.Width - 10) - (COLOR_PADDING * 2), this.Height - (COLOR_PADDING * 2));
+
+            colorBoxes = new Dictionary<Gw2Sharp.WebApi.V2.Models.Color, ColorBox>();
+        }
+
+        private void Colors_CollectionChanged(object sender, System.Collections.Specialized.NotifyCollectionChangedEventArgs e) {
+            // Remove any items that were removed (first because moving an item in a collection will put
+            // that item in both OldItems and NewItems)
+            if (e.OldItems != null) {
+                foreach (var deletedItem in e
+                                       .OldItems
+                                       .Cast<Gw2Sharp.WebApi.V2.Models.Color>()
+                                       .Where(delItem => colorBoxes.ContainsKey(delItem))) {
+                    colorBoxes[deletedItem].Dispose();
+                    colorBoxes.Remove(deletedItem);
+                }
+            }
+
+            if (e.NewItems != null) {
+                foreach (Gw2Sharp.WebApi.V2.Models.Color addedItem in e.NewItems) {
+                    if (!colorBoxes.ContainsKey(addedItem)) {
+                        var colorBox = new ColorBox() {
+                            Color = addedItem,
+                            Parent = this,
+                            Size = new Point(COLOR_SIZE),
+                        };
+
+                        colorBoxes[addedItem] = colorBox;
+
+                        colorBox.LeftMouseButtonPressed += delegate {
+                            foreach (var box in colorBoxes.Values) {
+                                box.Selected = false;
+                            }
+
+                            colorBox.Selected = true;
+                            this.SelectedColor = colorBox.Color;
+                        };
+                    }
+                }
+            }
+
+            // Relayout the color grid
+            for (int i = 0; i < this.Colors.Count; i++) {
+                var currentColor = this.Colors[i];
+                var currentBox = colorBoxes[currentColor];
+
+                int horizontalPosition = i % colorsPerRow;
+                int verticalPosition = i / colorsPerRow;
+
+                currentBox.Location = new Point(
+                    horizontalPosition * (currentBox.Width + COLOR_PADDING),
+                    verticalPosition * (currentBox.Width + COLOR_PADDING)
+                );
+            }
+        }
+
+        public override void Invalidate() {
```

Comment:
I believe this can get replaced with an override to `RecalculateLayout()`.  Nearly the same thing, but was introduced to let `Invalidate` handle some internal state stuff.  Technically fine since `Invalidate()` is called below.